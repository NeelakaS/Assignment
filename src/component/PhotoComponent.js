import React, {useEffect, useState} from 'react';
import {
  View,
  Text,
  FlatList,
  StyleSheet,
  TouchableOpacity,
  Image,
  ActivityIndicator,
} from 'react-native';
import {connect} from 'react-redux';
import DisplayImageModal from './modal/displayImageModal';
import {fetchPhotoData} from '../action/actions';
import styles from './PhotoComponent.style';

type PhotData = {
  albumId: number,
  id: number,
  title: string,
  url: string,
  thumbnailUrl: string,
};

type Props = {
  fetchPhotoInfo: (albumId: stirng) => void,
  photoData: PhotData,
  navigation: Objct,
  loading: boolean,
};

const PhotoComponent = (props: Props) => {
  const {fetchPhotoInfo, photoData, navigation, loading} = props;

  //hooks
  const [displyModal, setDisplayModal] = useState(false);
  const [url, setUrl] = useState('');
  const albumId = navigation.getParam('albumId');

  //hooks
  useEffect(() => {
    fetchPhotoInfo(albumId);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [albumId]);

  const showModal = (item: PhotData) => {
    setUrl(item.url);
    setDisplayModal(true);
  };
  const closeModal = () => {
    setDisplayModal(false);
  };

  const extractKey = (item: PhotData, index: number) => `${item?.id}-${index}`;

  const renderItem = ({item}: PhotData) => {
    return (
      <TouchableOpacity
        style={styles.listContainer}
        onPress={() => {
          showModal(item);
        }}>
        <Text style={styles.albumTitle}>{item.title}</Text>
        <Image
          source={{uri: item.thumbnailUrl}}
          style={styles.thumbnailImage}
        />
      </TouchableOpacity>
    );
  };
  // loading untill API receives data
  if (loading) {
    return (
      <View style={styles.actionIndicator}>
        <ActivityIndicator />
      </View>
    );
  }

  return (
    <View>
      <FlatList
        ItemSeparatorComponent={() => <View style={styles.listItemSeparator} />}
        data={photoData}
        renderItem={renderItem}
        keyExtractor={extractKey}
      />
      {/* Display Full size Image Modal */}
      {!!displyModal && (
        <DisplayImageModal
          visible={displyModal}
          closeModal={closeModal}
          imageUrl={url}
        />
      )}
    </View>
  );
};

const mapDispatchToProps = dispatch => ({
  fetchPhotoInfo: albumId => dispatch(fetchPhotoData(albumId)),
});
const mapStateToProps = state => ({
  photoData: state.photoData.data,
  loading: state.photoData.isFetching,
});

// eslint-disable-next-line prettier/prettier
export default connect(mapStateToProps, mapDispatchToProps)(PhotoComponent);
