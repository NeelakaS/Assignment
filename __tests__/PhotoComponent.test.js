import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import saga from 'redux-saga';

import PhotoComponent from '../src/component/PhotoComponent';

const PHOTO_DATA = [
  {
    albumId: 5,
    id: 215,
    title: 'in magnam praesentium ab illum',
    url: 'XYZ.Com/src',
    thumbnailUrl: 'www.yze.com',
  },
  {
    albumId: 5,
    id: 216,
    title: 'itaque nihil voluptatum',
    url: 'XYZ.Com/src',
    thumbnailUrl: 'www.yze.com',
  },
  {
    albumId: 5,
    id: 217,
    title: 'ab ut placeat fuga',
    url: 'XYZ.Com/src',
    thumbnailUrl: 'www.yze.com',
  },
];

const DEFAULT_PROPS = {
  fetchUserData: jest.fn(),
  photoData: [],
  navigation: {
    getParam: jest.fn(),
  },
  loading: false,
};

const middlewares = [saga];
const mockStore = configureMockStore(middlewares);
let wrapper;

describe('Photo Component', () => {
  it('render UserComponent correctly', () => {
    const initialState = {
      photoData: {
        data: PHOTO_DATA,
        isDataFetched: false,
        isFetching: false,
        isError: false,
      },
    };
    const store = mockStore(initialState);

    wrapper = renderer
      .create(
        <Provider store={store}>
          <PhotoComponent {...DEFAULT_PROPS} />
        </Provider>,
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
