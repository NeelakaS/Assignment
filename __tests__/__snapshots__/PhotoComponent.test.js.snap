// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo Component render UserComponent correctly 1`] = `
<View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "albumId": 5,
          "id": 215,
          "thumbnailUrl": "www.yze.com",
          "title": "in magnam praesentium ab illum",
          "url": "XYZ.Com/src",
        },
        Object {
          "albumId": 5,
          "id": 216,
          "thumbnailUrl": "www.yze.com",
          "title": "itaque nihil voluptatum",
          "url": "XYZ.Com/src",
        },
        Object {
          "albumId": 5,
          "id": 217,
          "thumbnailUrl": "www.yze.com",
          "title": "ab ut placeat fuga",
          "url": "XYZ.Com/src",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "height": 220,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 14,
                "fontWeight": "600",
                "lineHeight": 25,
              }
            }
          >
            in magnam praesentium ab illum
          </Text>
          <Image
            source={
              Object {
                "uri": "www.yze.com",
              }
            }
            style={
              Object {
                "height": 150,
                "width": 150,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#404041",
              "minHeight": 0.5,
              "opacity": 0.3,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "height": 220,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 14,
                "fontWeight": "600",
                "lineHeight": 25,
              }
            }
          >
            itaque nihil voluptatum
          </Text>
          <Image
            source={
              Object {
                "uri": "www.yze.com",
              }
            }
            style={
              Object {
                "height": 150,
                "width": 150,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#404041",
              "minHeight": 0.5,
              "opacity": 0.3,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "height": 220,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 14,
                "fontWeight": "600",
                "lineHeight": 25,
              }
            }
          >
            ab ut placeat fuga
          </Text>
          <Image
            source={
              Object {
                "uri": "www.yze.com",
              }
            }
            style={
              Object {
                "height": 150,
                "width": 150,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
